<a href="https://colab.research.google.com/github/odevhesabi1/Odevler/blob/main/%C3%96dev%20-%20Hedef%20De%C4%9Fi%C5%9Fken%20ve%20%C3%96zellikler%20Aras%C4%B1ndaki%20%C4%B0li%C5%9Fkiyi%20Anlama.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [7]:
import pandas as pd
import numpy as np
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model
import matplotlib.pyplot as plt

weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

In [8]:
weather_df["TemperatureDifference"] = weather_df["Apparent Temperature (C)"] - weather_df["Temperature (C)"]
weather_df.head()

#Hedef değişken negatif değerlere sahip, bu modeli etkiler mi ?
X = weather_df[["Humidity", "Wind Speed (km/h)"]]
Y = weather_df["TemperatureDifference"]

lrm = linear_model.LinearRegression()
lrm.fit(X,Y)

print('Değişkenler: \n', lrm.coef_)
print('Sabit değer (bias): \n', lrm.intercept_)

Değişkenler: 
 [-3.02918594 -0.11929075]
Sabit değer (bias): 
 2.4381054151877035


In [9]:
#Statsmodel ile yeni bir model oluşturma
import statsmodels.api as sm
X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()



<class 'statsmodels.iolib.summary.Summary'>
"""
                              OLS Regression Results                             
=================================================================================
Dep. Variable:     TemperatureDifference   R-squared:                       0.288
Model:                               OLS   Adj. R-squared:                  0.288
Method:                    Least Squares   F-statistic:                 1.949e+04
Date:                   Sat, 22 May 2021   Prob (F-statistic):               0.00
Time:                           12:53:12   Log-Likelihood:            -1.7046e+05
No. Observations:                  96453   AIC:                         3.409e+05
Df Residuals:                      96450   BIC:                         3.409e+05
Df Model:                              2                                         
Covariance Type:               nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 2.4381      0.021    115.948      0.000       2.397       2.479
Humidity             -3.0292      0.024   -126.479      0.000      -3.076      -2.982
Wind Speed (km/h)    -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Model katsayılarının P değerleri 0 olduğundan istatistiksel olarak anlamlıdır.( İstatistiksel olarak anlamlı olması tam olarak neyi ifade eder, neyi açıklar? )

Özelliklerin değerindeki artış, katsayılar negatif değer olduğundan hedef değerde azalış anlamına gelir. (Görünür sıcaklık farkı azalır demek).

Windspeed'in katsayısı sıfıra yakın (0.1), bu yüzden varyansın az kısmını taşıdığını (bilginin az kısmını taşıdığını düşünüyorum.)

Humidity'nin katsayısı daha yüksek olduğundan bilginin coğunluğunun bu özellikte olduğunu, varyansın bu özellikte daha çok olduğunu düşünüyorum

Bir önceki modelde R squared 0.44'tü hatırladığım kadarıyla, bu modelimizde R squared 0.288. Yani model daha az hataya sahip ve daha iyi bir model ?

In [10]:
#Ardından, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Katsayılar istatistiksel olarak anlamlı mıdır? 
#Nem ve rüzgar hızı için tahmini katsayıların işaretleri değişti mi? Tahmin edilen katsayıları yorumlayınız.

weather_df["Hum_and_Wind"] = weather_df["Humidity"] * weather_df["Wind Speed (km/h)"]

X = weather_df[["Humidity", "Wind Speed (km/h)", "Hum_and_Wind"]]
Y = weather_df["TemperatureDifference"]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                              OLS Regression Results                             
=================================================================================
Dep. Variable:     TemperatureDifference   R-squared:                       0.341
Model:                               OLS   Adj. R-squared:                  0.341
Method:                    Least Squares   F-statistic:                 1.666e+04
Date:                   Sat, 22 May 2021   Prob (F-statistic):               0.00
Time:                           12:53:12   Log-Likelihood:            -1.6669e+05
No. Observations:                  96453   AIC:                         3.334e+05
Df Residuals:                      96449   BIC:                         3.334e+05
Df Model:                              3                                         
Covariance Type:               nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.0839      0.033      2.511      0.012       0.018       0.149
Humidity              0.1775      0.043      4.133      0.000       0.093       0.262
Wind Speed (km/h)     0.0905      0.002     36.797      0.000       0.086       0.095
Hum_and_Wind         -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Yorumlarımı yaptım ama katsayıların yorumlanmasını yinede beraber yapabilir miyiz.

Sabitimizin P değeri artmış görünüyor, ancak yinede 0.1'den küçük olduğundan istatistiksel olarak hala anlamlı.

Nem ve rüzgar özelliklerinin bireysel katsayısı azalmış, yeni oluşturduğumuz etkileşimli özelliğin katsayısı ise bunlara göre oldukça yüksek görünüyor. Hedef değişkenindeki varyansın çoğunu bu yeni özellik açıklıyor. ( Yorumum doğru ise final modeli oluştururken nem ve rüzgar hızı özelliğini çıkarıp, modele sadece bunların etkileşimli yeni oluşturduğumuz özelliğini ekleyerek mi devam etmek gerekir ?)

Etkileşimli özelliğin P değeri 0 olduğundan istatistiksel olarak anlamlı.( Sadece p değerine bakarak eğer p değeri yüksekse o zaman o özelliğin katsayısı sıfıra yakın oldugundan, model üzerinde aslında pek bir etkiye sahip olmadığını düşünüp başka birşeye bakmadan, direkt modelden cıkarabilir miyiz ? )

Ayrıca Rsquared yani modelin hatası artmış.( istatistiği doğru mu okuyorum ?) Bu yüzden bir önceki modelin Rsqaure'i daha düşük oldugundan o daha iyi bir modeldi.

In [33]:
#  3. Ev fiyatları modeli

house_prices_df1 = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df1.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [12]:
house_prices_df.isnull().sum() / house_prices_df.shape[0]

Id               0.000000
MSSubClass       0.000000
MSZoning         0.000000
LotFrontage      0.177397
LotArea          0.000000
                   ...   
MoSold           0.000000
YrSold           0.000000
SaleType         0.000000
SaleCondition    0.000000
SalePrice        0.000000
Length: 81, dtype: float64

In [35]:
house_prices_df2 = house_prices_df1.copy()
house_prices_df2 = house_prices_df1.drop(["Alley", "FireplaceQu", "PoolQC", "Fence", "MiscFeature"], inplace=True, axis=1 )
house_prices_df2.head()

KeyError: ignored

Ev fiyatları için model olusturmadan önce, nerdeyse tamamen null olan bazı kolonları direkt drop etmeye calıstım ancak yukarıdaki hatayı verdi. (Kategorik değerleride içeren, büyük bir veri. Nasıl temizlememiz gerektiğine hızlıca beraber bakabilir miyiz ? )

Aşağıda model için korelasyon katsayılarına bakıp, hedef özellik ile korelasyonu yüksek birkaç özelliği alarak modeli inşaa edeceğim. ( Diyelim veriseti hiç bir bilgi sahibi olmadığımız bir alan ile ilgili, özellikler ve hedef değişken ile ilişkisi hakkında hiçbir fikrimiz yok. Bu durumda feature engineering yaparken ilk olarak korelasyon matrisine bakıp, korelasyonu yüksek özellikleri alarak yola devam etme fikrine sahibim, bu yanlış mı ? Doğru bir yaklaşım ise seçeceğimiz özelliklerin hepsi 0.8'den düşük olmak şartıyla yüksek korelasyonlu özellikler mi olmalı yoksa düşük korelasyonları da içeren bir kombinasyon mu yapmalıyız ? )

In [ ]:
house_prices_df1.corr()

Ev fiyatları veri setini daha öncede temizleyemediğim için yine yapamıyorum.

1- Kategorik değişkenler içeriyor nasıl temizlemeliyiz.

2-Veriseti çok büyük, modele geçmeden önce aykırı değer temizliği yapmak gerekiyor. Çok özellik içeren verisetlerinde bunu nasıl yapacağız, görselleştirme vs kullanmak özellikler çok olduğundan cok fazla zaman alır. O yüzden direkt winsorize mi etmek gerekir? Öyleyse görselleştirmeden yine winsorizenin aşağı-yukarı yönlü ya da winsorize parametresi olarak ne vereceğimi nasıl bilebilirim ? 80 tane feature için bunları incelemek çok fazla zaman alıyor ?)

3- Kullanmayacağım özellikler için yukarıdaki veri temizleme, winsorize, normalleştirme gibi işlemleri yapmak zaman kaybı gibi geliyor. Direkt korelasyona bakarak featureları seçip ondan sonra bu işlemleri yapmak mantıklı geliyor.

Ama aykırı değerleri temizlemeden, normalleştirme vs. gibi işlemleri uygulamadan korelasyona bakmak mantıklı mıdır. Bunları yapıp korelasyona baktıktan sonra korelasyon sonucları farklı mı cıkar ?